source/server/infrastructure/projection-rebuilder.coffee

```diff
@@ -17,39 +17,66 @@ class Space.eventSourcing.ProjectionRebuilder extends Space.Object
 
     realCollectionsBackups = {}
     queue = []
+    startHrTime = process.hrtime()
 
     # Loop over all projections that should be rebuilt
     for projectionId in projections
       projection = @injector.get projectionId
+
+      # Tell the projection that it will be rebuilt now
+      try
+        projection.enterRebuildMode()
+      catch error
+        if error instanceof Space.eventSourcing.ProjectionAlreadyRebuilding
+          @log.warning(@_logMsg(error.message))
+          return { error: error }
+        else
+          throw error
+
+      @log.info(@_logMsg("Rebuilding #{projection}"))
+
       # Save backups of the real collections to restore them later and
       # override the real collections with in-memory pendants
       for collectionId in @_getCollectionIdsOfProjection(projection)
         realCollectionsBackups[collectionId] = @injector.get collectionId
+        @log.debug(@_logMsg("Backed up #{collectionId}"))
         @injector.override(collectionId).to new @mongo.Collection(null)
+        @log.debug(@_logMsg("Injector mapping for #{collectionId} overridden with in-memory staging collection"))
 
-      # Tell the projection that it will be rebuilt now
-      projection.enterRebuildMode()
       queue.push projection
 
-    # Loop through all events and hand them individually to all projections
-    for event in @commitStore.getAllEvents()
-      projection.on(event, true) for projection in queue
-
-    # Update the real collection data with the in-memory versions
-    # for the specified projections only.
-    for collectionId, realCollection of realCollectionsBackups
-      inMemoryCollection = @injector.get(collectionId)
-      inMemoryData = inMemoryCollection.find().fetch()
-      realCollection.remove {}
-      if inMemoryData.length
-        realCollection.batchInsert inMemoryData
-      else
-        @log.info(@_logMsg("No data to insert after replaying events for #{collectionId}"))
-      # Restore original collections
-      @injector.override(collectionId).to realCollection
+    try
+      # Loop through all events and hand them individually to all projections
+      @log.debug(@_logMsg("Starting to pass events to #{projectionId} from Commit Store"))
+      for event in @commitStore.getAllEvents()
+        projection.on(event, true) for projection in queue
+      @log.debug(@_logMsg("Finished passing events to #{projectionId}"))
```

Comment:
This log message is misleading as we are passing each event to all projections in the queue! We are not rebuilding one collection at the time but all together, step by step (event history) as this is necessary if they are cross-referencing each other!